### PR TITLE
[GPU][NFC] Further simplify creation of cuDNN FMHA graphs.

### PR DESCRIPTION
**Pull request by sergachev**
This is a follow-up to https://github.com/openxla/xla/pull/15919.

### CODE REVIEW

**Review by ezhulenev — APPROVED** [review]

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc [hunk 1]
```diff
@@ -541,47 +246,134 @@ absl::StatusOr<se::gpu::CudnnGraph> HloCustomCallToCuDnnGraph(
...
+        bmm1_grad_gemm2_dnums.rhs_batch_dimensions(),
+        bmm1_grad_gemm2_dnums.rhs_contracting_dimensions());
+
+    MatmulTensorDescriptor bmm2_grad_gemm1_lhs = MatmulTensorDescriptor::For(
+        bmm2_grad_gemm1_lhs_type, bmm2_grad_gemm1_lhs_shape.dimensions(),
+        bmm2_grad_gemm1_lhs_shape.layout().minor_to_major(),
+        bmm2_grad_gemm1_dnums.lhs_batch_dimensions(),
+        bmm2_grad_gemm1_dnums.lhs_contracting_dimensions());
+
+    MatmulTensorDescriptor d_output = MatmulTensorDescriptor::For(
```
Comment (jreiffers) [inline xla/service/gpu/transforms/cudnn_custom_call_compiler.cc:301]: Nit: maybe this should come after bmm2_grad_gemm2_rhs?

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc [hunk 2]
```diff
@@ -541,47 +246,134 @@ absl::StatusOr<se::gpu::CudnnGraph> HloCustomCallToCuDnnGraph(
     config.set_force_deterministic(force_deterministic);
     TF_RETURN_IF_ERROR(custom_call->set_backend_config(gpu_config));
```
Comment (jreiffers) [inline xla/service/gpu/transforms/cudnn_custom_call_compiler.cc:248]: There's a lot of repetition here. How about putting some of these locals into a struct?

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc [hunk 3]
```diff
@@ -541,47 +246,134 @@ absl::StatusOr<se::gpu::CudnnGraph> HloCustomCallToCuDnnGraph(
...
+      mask = TensorDescriptor::For(mask_type, mask_shape->dimensions(),
+                                   mask_shape->layout().minor_to_major());
+    }
+
+    std::optional<se::dnn::TensorDescriptor> fwd_output;
+    if (fwd_output_shape.has_value()) {
+      TF_ASSIGN_OR_RETURN(
+          DataType fwd_output_type,
+          GetDNNDataTypeFromPrimitiveType(fwd_output_shape->element_type()));
+      fwd_output =
```
Comment (jreiffers) [inline xla/service/gpu/transforms/cudnn_custom_call_compiler.cc:347]: Did you forget to use all these optionals or am I missing something?

**Comment by jreiffers** [comment]
FYI this was accidentally merged and then rolled back.

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc (hunk 1 above)
Comment (sergachev) [reply xla/service/gpu/transforms/cudnn_custom_call_compiler.cc]: Done.

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc (hunk 2 above)
Comment (sergachev) [reply xla/service/gpu/transforms/cudnn_custom_call_compiler.cc:223]: Added TensorDescriptorFor() and MatmulTensorDescriptorFor() which remove lots of repeated code.

File: xla/service/gpu/transforms/cudnn_custom_call_compiler.cc (hunk 3 above)
Comment (sergachev) [reply xla/service/gpu/transforms/cudnn_custom_call_compiler.cc]: This means that they were not in use already before I started simplifying. Removed a few more.

**Review by jreiffers — APPROVED** [review]
Thanks!

**Pull request closed without merging**